File: apps/backend/src/EnterpriseAgentOs.Domain/Common/ValueObjects/CreditBudget.cs
namespace EnterpriseAgentOs.Domain.Common.ValueObjects;

/// <summary>
/// Encapsulates credit budget state and logic shared between UserSubscription and OrgSubscription.
/// </summary>
public sealed class CreditBudget
{
    public long BudgetPerMonth { get; set; }
    public long UsedThisMonth { get; set; }
    public bool OverageEnabled { get; set; }

    public long Remaining => BudgetPerMonth - UsedThisMonth;
    public bool IsOverBudget => Remaining < 0;

    public CreditBudget(long budgetPerMonth, long usedThisMonth, bool overageEnabled)
    {
        BudgetPerMonth = budgetPerMonth;
        UsedThisMonth = usedThisMonth;
        OverageEnabled = overageEnabled;
    }

    public CreditBudgetResult Check() => new(Remaining, IsOverBudget);

    public void Record(long credits) => UsedThisMonth += credits;
}

File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/common/valueobjects/CreditBudget.java
package com.enterpriseagentos.backendjava.domain.common.valueobjects;

import com.enterpriseagentos.backendjava.domain.features.management.dtos.CreditBudgetResult;

public class CreditBudget {
    private long budgetPerMonth;
    private long usedThisMonth;
    private boolean overageEnabled;

    public CreditBudget(long budgetPerMonth, long usedThisMonth, boolean overageEnabled) {
        this.budgetPerMonth = budgetPerMonth;
        this.usedThisMonth = usedThisMonth;
        this.overageEnabled = overageEnabled;
    }

    public long remaining() {
        return budgetPerMonth - usedThisMonth;
    }

    public boolean isOverBudget() {
        return remaining() < 0;
    }

    public CreditBudgetResult check() {
        return new CreditBudgetResult(remaining(), isOverBudget());
    }

    public void addCredits(long credits) {
        usedThisMonth += credits;
    }

    public long getBudgetPerMonth() {
        return budgetPerMonth;
    }

    public long getUsedThisMonth() {
        return usedThisMonth;
    }

    public boolean getOverageEnabled() {
        return overageEnabled;
    }
}
